Cabal/Distribution/PackageDescription/Parsec.hs

```diff
@@ -0,0 +1,566 @@
+            -- TODO check duplicate name here?
+            let gpd' = gpd { condBenchmarks = condBenchmarks gpd ++ [(name', bench)] }
+            pure gpd'
+
+        | name == "flag" = do
+            name' <- parseName pos args
+            name'' <- runFieldParser' pos parsec name' `recoverWith` FlagName ""
+            flag <- parseFields flagFieldDescrs warnUnrec (emptyFlag name'') fields
+            -- Check default flag
+            let gpd' = gpd { genPackageFlags = genPackageFlags gpd ++ [flag] }
+            pure gpd'
+
+        | name == "custom-setup" && null args = do
+            sbi <- parseFields setupBInfoFieldDescrs warnUnrec mempty fields
+            let pd = packageDescription gpd
+            -- TODO: what if already defined?
+            let gpd' = gpd { packageDescription = pd { setupBuildInfo = Just sbi } }
+            pure gpd'
+
+        | name == "source-repository" = do
+            kind <- case args of
+                [SecArgName spos secName] ->
+                    runFieldParser' spos parsec (fromUTF8BS secName) `recoverWith` RepoHead
+                [] -> do
+                    parseFailure pos $ "'source-repository' needs one argument"
+                    pure RepoHead
+                _ -> do
+                    parseFailure pos $ "Invalid source-repository kind " ++ show args
+                    pure RepoHead
+            sr <- parseFields sourceRepoFieldDescrs warnUnrec (emptySourceRepo kind) fields
+            -- I want lens
+            let pd =  packageDescription gpd
+            let srs = sourceRepos pd
+            let gpd' = gpd { packageDescription = pd { sourceRepos = srs ++ [sr] } }
+            pure gpd'
+
+        | otherwise = do
+            parseWarning pos PWTUnknownSection $ "Ignoring section: " ++ show name
+            pure gpd
+
+    newSyntaxVersion :: Version
+    newSyntaxVersion = mkVersion [1, 2]
+
+    maybeWarnCabalVersion :: Syntax -> PackageDescription -> ParseResult ()
+    maybeWarnCabalVersion syntax pkg
+      | syntax == NewSyntax && specVersion pkg < newSyntaxVersion
+      = parseWarning (Position 0 0) PWTNewSyntax $
+             "A package using section syntax must specify at least\n"
+          ++ "'cabal-version: >= 1.2'."
+
+    maybeWarnCabalVersion syntax pkg
+      | syntax == OldSyntax && specVersion pkg >= newSyntaxVersion
+      = parseWarning (Position 0 0) PWTOldSyntax $
+             "A package using 'cabal-version: "
+          ++ displaySpecVersion (specVersionRaw pkg)
+          ++ "' must use section syntax. See the Cabal user guide for details."
+      where
+        displaySpecVersion (Left version)       = display version
+        displaySpecVersion (Right versionRange) =
+          case asVersionIntervals versionRange of
+            [] {- impossible -}           -> display versionRange
+            ((LowerBound version _, _):_) -> display (orLaterVersion version)
+
+    maybeWarnCabalVersion _ _ = return ()
+
+{-
```

Comment:
they are "TODOs", there seems to not be tests for those, so didn't remember to implement